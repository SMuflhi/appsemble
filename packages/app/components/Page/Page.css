.error {
  margin: 20vh 8px 8px;
}

.dotContainer {
  padding-right: 15px;
  padding-top: 15px;
  position: absolute;
  right: 0;
  z-index: 1000;
}

.dot {
  background-color: hsl(0, 0%, 100%);
  border: none;
  border-radius: 50%;
  display: inline-block;
  height: 10px;
  margin: 0 2px;
  opacity: 0.2;
  width: 10px;
}

.dot.previous {
  opacity: 0.5;
}

.dot.active {
  opacity: 1;
}

.transitionGroup,
.transitionWrapper {
  display: flex;
  height: 100%;
  position: relative;
}

.transitionWrapper {
  left: 0;
  position: absolute;
  top: 0;
  width: 100%;
}

.pageEnter {
  opacity: 0;
}

.pageEnterActive {
  opacity: 1;
  transition: opacity 300ms ease-in;
}

.pageExit {
  opacity: 1;
}

.pageExitActive {
  opacity: 0;
  transition: opacity 300ms ease-in;
}
