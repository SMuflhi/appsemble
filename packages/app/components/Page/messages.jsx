import { defineMessages } from 'react-intl';

export default defineMessages({
  error: 'There was a problem loading this page. Weâ€™re sorry.',
  permissionLogout:
    'You have been logged out due to not having sufficient permissions. Please contact the app owner if you think this is a mistake.',
});
