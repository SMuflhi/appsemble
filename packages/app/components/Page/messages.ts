import { defineMessages } from 'react-intl';

export const messages = defineMessages({
  permissionError: {
    id: 'app.y/1N+5',
    defaultMessage:
      'Due to misconfigured permissions, youâ€™re not able to view the app. If you think this is a mistake, please leave a message in the <link>Appsemble store page</link>.',
  },
  logout: {
    id: 'app.C81/uG',
    defaultMessage: 'Logout',
  },
});
