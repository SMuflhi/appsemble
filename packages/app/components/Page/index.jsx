import { connect } from 'react-redux';

import { getBlockDefs } from '../../actions/blockDefs';
import Page from './Page';

function mapStateToProps(state, ownProps) {
  return {
    app: state.app.app,
    user: state.user.user,
    hasErrors: ownProps.page.blocks.some(block => state.blockDefs.errored.has(block.type)),
  };
}

export default connect(
  mapStateToProps,
  {
    getBlockDefs,
  },
)(Page);
