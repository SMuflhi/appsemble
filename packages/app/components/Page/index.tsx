import { connect } from 'react-redux';

import { State } from '../../actions';
import { getBlockDefs } from '../../actions/blockDefs';
import { logout } from '../../actions/user';
import Page from './Page';

function mapStateToProps(
  state: State,
  ownProps: Pick<React.ComponentPropsWithoutRef<typeof Page>, 'page'>,
): Partial<React.ComponentPropsWithoutRef<typeof Page>> {
  return {
    definition: state.app.definition,
    user: state.user.user,
    role: state.user.role,
    pending: !!state.blockDefs.pending.length,
    hasErrors:
      ownProps.page.type && (ownProps.page.type === 'flow' || ownProps.page.type === 'tabs')
        ? ownProps.page.subPages.some(sub =>
            sub.blocks.some(block => state.blockDefs.errored.has(block.type)),
          )
        : ownProps.page.blocks.some(block => state.blockDefs.errored.has(block.type)),
  };
}

export default connect(mapStateToProps, {
  getBlockDefs,
  logout,
})(Page);
