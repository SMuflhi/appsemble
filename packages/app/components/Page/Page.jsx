import { normalize } from '@appsemble/utils';
import classNames from 'classnames';
import EventEmitter from 'events';
import throttle from 'lodash.throttle';
import PropTypes from 'prop-types';
import React from 'react';
import { FormattedMessage } from 'react-intl';
import { Link, Redirect, Route, Switch } from 'react-router-dom';
import { CSSTransition, TransitionGroup } from 'react-transition-group';

import checkScope from '../../utils/checkScope';
import makeActions from '../../utils/makeActions';
import Block from '../Block';
import Login from '../Login';
import PageDialog from '../PageDialog';
import TitleBar from '../TitleBar';
import messages from './messages';
import styles from './Page.css';

/**
 * Render an app page definition.
 */
export default class Page extends React.Component {
  static propTypes = {
    app: PropTypes.shape().isRequired,
    getBlockDefs: PropTypes.func.isRequired,
    hasErrors: PropTypes.bool.isRequired,
    history: PropTypes.shape().isRequired,
    match: PropTypes.shape().isRequired,
    /**
     * The page definition to render
     */
    page: PropTypes.shape().isRequired,
    user: PropTypes.shape(),
  };

  static defaultProps = {
    user: null,
  };

  state = {
    dialog: null,
    actions: {},
    counter: 0,
    currentPage: 0,
    data: {},
  };

  flowActions = {
    next: throttle(
      async data => {
        const { currentPage } = this.state;
        const { page } = this.props;
        const { subPages } = page;

        if (currentPage + 1 === subPages.length) {
          this.actions.onFlowFinish.dispatch(data);
          return data;
        }

        this.setState({ data, currentPage: currentPage + 1 });
        return data;
      },
      50,
      { leading: false },
    ),

    finish: throttle(
      async data => {
        this.actions.onFlowFinish.dispatch(data);
        this.setState({ data });
        return data;
      },
      50,
      { leading: false },
    ),

    back: throttle(
      async data => {
        const { currentPage } = this.state;

        if (currentPage <= 0) {
          // Don't do anything if a previous page does not exist
          return data;
        }

        this.setState({ data, currentPage: currentPage - 1 });
        return data;
      },
      50,
      { leading: false },
    ),

    cancel: throttle(
      async data => {
        this.actions.onFlowCancel.dispatch(data);
        this.setState({ data });
      },
      50,
      { leading: false },
    ),
  };

  constructor(props) {
    super(props);
    this.setupEvents();
  }

  componentDidMount() {
    const { app, getBlockDefs, page, history } = this.props;

    this.applyBulmaThemes(app, page);
    this.setupEvents();

    if (page.type === 'tabs' || page.type === 'flow') {
      if (page.type === 'flow') {
        const actions = makeActions(
          { actions: { onFlowFinish: {}, onFlowCancel: {} } },
          app,
          page,
          history,
          this.showDialog,
          {
            emit: this.emitEvent,
            off: this.offEvent,
            on: this.onEvent,
          },
          {},
          this.flowActions,
        );

        this.actions = actions;
      }

      getBlockDefs(
        [
          ...new Set(
            page.subPages
              .map(f => f.blocks)
              .flat()
              .map(b => JSON.stringify({ type: b.type, version: b.version })),
          ),
        ].map(block => JSON.parse(block)),
      );
    } else {
      getBlockDefs(page.blocks);
    }
  }

  static getDerivedStateFromProps(props, state) {
    if (state.prevPage !== props.page) {
      return { ...state, prevPage: props.page, counter: state.counter + 1 };
    }

    // Nothing to update.
    return null;
  }

  componentDidUpdate({ page: prevPage }, { prevCurrentPage }) {
    const { app, getBlockDefs, page } = this.props;
    const { currentPage } = this.state;

    if (page !== prevPage || prevCurrentPage !== currentPage) {
      this.teardownEvents();
      this.setupEvents();

      if (page.type === 'flow' || page.type === 'tabs') {
        getBlockDefs(
          [
            ...new Set(
              page.subPages
                .map(f => f.blocks)
                .flat()
                .map(b => JSON.stringify({ type: b.type, version: b.version })),
            ),
          ].map(block => JSON.parse(block)),
        );
      } else {
        getBlockDefs(page.blocks);
      }

      this.applyBulmaThemes(app, page);
    }
  }

  componentWillUnmount() {
    this.teardownEvents();
  }

  setupEvents() {
    const ee = new EventEmitter();
    this.emitEvent = (name, data) => ee.emit(name, data);
    this.offEvent = (name, callback) => ee.off(name, callback.bind());
    this.onEvent = (name, callback) => ee.on(name, callback.bind());
    this.ee = ee;
  }

  createBulmaQueryString = () => {
    const { app, page } = this.props;
    const params = { ...app.theme, ...page.theme };
    const queryStringParams = new URLSearchParams(params);
    queryStringParams.sort();

    return queryStringParams.toString();
  };

  applyBulmaThemes = (app, page) => {
    const bulmaStyle = document.getElementById('bulma-style-app');
    const [bulmaUrl] = bulmaStyle.href.split('?');
    bulmaStyle.href =
      app.theme || page.theme ? `${bulmaUrl}?${this.createBulmaQueryString()}` : bulmaUrl;
  };

  showDialog = dialog => {
    this.setState({ dialog });
    return () => {
      this.setState({ dialog: null });
    };
  };

  teardownEvents() {
    if (this.ee) {
      this.ee.removeAllListeners();
      this.ee = null;
      this.emitEvent = null;
      this.offEvent = null;
      this.onEvent = null;
    }
  }

  render() {
    const { hasErrors, page, user, match } = this.props;
    const { dialog, counter, currentPage, data } = this.state;
    const { type } = page;

    if (!checkScope(page.scope, user)) {
      return (
        <React.Fragment>
          <TitleBar>{page.name}</TitleBar>
          <Login />
        </React.Fragment>
      );
    }

    if (hasErrors) {
      return (
        <p className={styles.error}>
          <FormattedMessage {...messages.error} />
        </p>
      );
    }

    switch (type) {
      case 'flow':
        return (
          <React.Fragment>
            <TitleBar>{page.name}</TitleBar>
            <div className={styles.dotContainer}>
              {page.subPages.map((sub, index) => (
                <div
                  key={sub.name}
                  className={`${styles.dot} ${index < currentPage && styles.previous} ${index ===
                    currentPage && styles.active}`}
                />
              ))}
            </div>
            <TransitionGroup className={styles.transitionGroup}>
              {page.subPages[currentPage].blocks.map((block, index) => (
                <CSSTransition
                  // Since blocks are in a static list, using the index as a key should be fine.
                  // eslint-disable-next-line react/no-array-index-key
                  key={`${currentPage}.${index}.${counter}`}
                  classNames={{
                    enter: styles.pageEnter,
                    enterActive: styles.pageEnterActive,
                    exit: styles.pageExit,
                    exitActive: styles.pageExitActive,
                  }}
                  timeout={300}
                >
                  <div className={styles.transitionWrapper}>
                    <Block
                      block={block}
                      className="foo"
                      data={data}
                      emitEvent={this.emitEvent}
                      flowActions={this.flowActions}
                      offEvent={this.offEvent}
                      onEvent={this.onEvent}
                      showDialog={this.showDialog}
                    />
                  </div>
                </CSSTransition>
              ))}
            </TransitionGroup>
            <PageDialog
              dialog={dialog}
              emitEvent={this.emitEvent}
              offEvent={this.offEvent}
              onEvent={this.onEvent}
            />
          </React.Fragment>
        );
      case 'tabs':
        return (
          <React.Fragment>
            <TitleBar>{page.name}</TitleBar>
            <div className="tabs is-centered is-medium">
              <ul>
                {page.subPages.map(({ name }) => (
                  <li
                    key={name}
                    className={classNames({
                      'is-active': normalize(name) === match.params.subPage,
                    })}
                  >
                    <Link to={`${normalize(name)}`}>{name}</Link>
                  </li>
                ))}
              </ul>
            </div>
            <Switch>
              {page.subPages.map(({ name, blocks }) => (
                <Route
                  key={name}
                  exact
                  path={`${match.path}/${normalize(name)}`}
                  render={() =>
                    blocks.map((block, index) => (
                      <Block
                        // eslint-disable-next-line react/no-array-index-key
                        key={`${name}.${index}.${counter}`}
                        block={block}
                        className="foo"
                        data={data}
                        emitEvent={this.emitEvent}
                        flowActions={this.flowActions}
                        offEvent={this.offEvent}
                        onEvent={this.onEvent}
                        showDialog={this.showDialog}
                      />
                    ))
                  }
                />
              ))}

              <Redirect to={`${match.url}/${normalize(page.subPages[0].name)}`} />
            </Switch>
          </React.Fragment>
        );
      case 'page':
      case 'subPage':
      default:
        return (
          <React.Fragment>
            {type !== 'subPage' && <TitleBar>{page.name}</TitleBar>}
            {page.blocks.map((block, index) => (
              <Block
                // As long as blocks are in a static list, using the index as a key should be fine.
                // eslint-disable-next-line react/no-array-index-key
                key={`${index}.${counter}`}
                block={block}
                emitEvent={this.emitEvent}
                flowActions={this.flowActions}
                offEvent={this.offEvent}
                onEvent={this.onEvent}
                showDialog={this.showDialog}
              />
            ))}
            <PageDialog
              dialog={dialog}
              emitEvent={this.emitEvent}
              offEvent={this.offEvent}
              onEvent={this.onEvent}
            />
          </React.Fragment>
        );
    }
  }
}
