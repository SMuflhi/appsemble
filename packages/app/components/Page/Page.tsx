import { Loader } from '@appsemble/react-components';
import useMessages from '@appsemble/react-components/hooks/useMessages';
import {
  ActionDefinition,
  AppDefinition,
  Block,
  DialogActionDefinition,
  Page as PageType,
} from '@appsemble/types';
import { normalize } from '@appsemble/utils';
import EventEmitter from 'events';
import React from 'react';
import { FormattedMessage, useIntl } from 'react-intl';
import { useHistory } from 'react-router-dom';

import { ShowDialogParams, User } from '../../types';
import checkAppRole from '../../utils/checkAppRole';
import BlockList from '../BlockList';
import FlowPage from '../FlowPage';
import Login from '../Login';
import PageDialog from '../PageDialog';
import TabsPage from '../TabsPage';
import TitleBar from '../TitleBar';
import messages from './messages';
import styles from './Page.css';

interface PageProps {
  definition: AppDefinition;
  role: string;
  getBlockDefs: (blocks: Block[]) => void;
  hasErrors: boolean;
  pending: boolean;
  page: PageType;
  user: User;
  logout: () => void;
}

export default function Page({
  definition,
  role,
  getBlockDefs,
  hasErrors,
  pending,
  page,
  user,
  logout,
}: PageProps): React.ReactElement {
  const history = useHistory();
  const intl = useIntl();
  const push = useMessages();

  const [dialog, setDialog] = React.useState<ShowDialogParams>();
  const [blocks, setBlocks] = React.useState<Block[]>([]);
  const [counter, setCounter] = React.useState(0);

  const ee = React.useRef<EventEmitter>();

  const createBulmaQueryString = React.useCallback(() => {
    const params = { ...definition.theme, ...page.theme };
    const queryStringParams = new URLSearchParams(params);
    queryStringParams.sort();

    return queryStringParams.toString();
  }, [definition, page.theme]);

  const applyBulmaThemes = React.useCallback(
    (d: AppDefinition, p: PageType) => {
      const bulmaStyle = document.getElementById('bulma-style-app') as HTMLLinkElement;
      const [bulmaUrl] = bulmaStyle.href.split('?');
      bulmaStyle.href = d.theme || p.theme ? `${bulmaUrl}?${createBulmaQueryString()}` : bulmaUrl;
    },
    [createBulmaQueryString],
  );

  const checkPagePermissions = React.useCallback(
    (p: PageType): boolean => {
      const roles = p.roles || definition.roles || [];
      return roles.length === 0 || roles.some(r => checkAppRole(definition.security, r, role));
    },
    [definition.roles, definition.security, role],
  );

  const handlePagePermissions = React.useCallback((): void => {
    const permission = checkPagePermissions(page);
    if (!permission) {
      const defaultPagePermission = checkPagePermissions(
        definition.pages.find(p => p.name === definition.defaultPage),
      );

      if (defaultPagePermission) {
        history.replace('/');
      } else {
        const redirectPage = definition.pages.find(
          p => p.parameters === undefined && checkPagePermissions(p),
        );

        if (!redirectPage) {
          push({
            body: intl.formatMessage(messages.permissionLogout),
            color: 'danger',
            dismissable: true,
          });
          logout();
          return;
        }

        history.replace(`/${normalize(redirectPage.name)}`);
      }
    }
  }, [checkPagePermissions, definition, history, intl, logout, page, push]);

  const showDialog = React.useCallback((d: ShowDialogParams) => {
    setDialog(d);
    return () => {
      setDialog(null);
    };
  }, []);

  React.useEffect(() => {
    setBlocks([
      ...(page.type === 'tabs' || page.type === 'flow'
        ? page.subPages.map(f => f.blocks).flat()
        : []),
      ...(!page.type || page.type === 'page' ? page.blocks : []),
    ]);
  }, [page]);

  React.useEffect(() => {
    applyBulmaThemes(definition, page);
    ee.current = new EventEmitter();

    const actionBlocks = blocks
      .filter(block => block.actions)
      .map(block =>
        Object.values<ActionDefinition>(block.actions)
          .filter(action => action.type === 'dialog')
          .map((action: DialogActionDefinition) => action.blocks),
      )
      .flat(2);

    getBlockDefs([...new Set([...blocks, ...actionBlocks])]);

    return () => {
      if (ee.current) {
        ee.current.removeAllListeners();
        ee.current = null;
      }
    };
  }, [applyBulmaThemes, blocks, definition, getBlockDefs, page]);

  const { type } = page;

  if (definition.security && !(page.roles && page.roles.length === 0)) {
    if (!user) {
      return (
        <>
          <TitleBar>{page.name}</TitleBar>
          <Login />
        </>
      );
    }

    handlePagePermissions();
  }

  if (hasErrors) {
    return (
      <p className={styles.error}>
        <FormattedMessage {...messages.error} />
      </p>
    );
  }

  if (pending) {
    return <Loader />;
  }

  let component;
  switch (type) {
    case 'flow':
      component = (
        <FlowPage
          blocks={blocks}
          counter={counter}
          definition={definition}
          ee={ee.current}
          page={page}
          showDialog={showDialog}
        />
      );
      break;
    case 'tabs':
      component = (
        <TabsPage
          counter={counter}
          ee={ee.current}
          showDialog={showDialog}
          subPages={page.subPages}
        />
      );
      break;
    default:
      component = (
        <BlockList blocks={page.blocks} counter={counter} ee={ee.current} showDialog={showDialog} />
      );
  }

  return (
    <>
      <TitleBar>{page.name}</TitleBar>
      {component}
      <PageDialog dialog={dialog} ee={ee.current} />
    </>
  );
}
