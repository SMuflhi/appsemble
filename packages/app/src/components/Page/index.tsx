import { Button, Content, Message, useLocationString } from '@appsemble/react-components';
import type { PageDefinition } from '@appsemble/types';
import { checkAppRole, normalize } from '@appsemble/utils';
import { EventEmitter } from 'events';
import React, { ReactElement, useCallback, useEffect, useRef, useState } from 'react';
import { FormattedMessage } from 'react-intl';
import { Redirect, Route, Switch, useRouteMatch } from 'react-router-dom';

import type { ShowDialogParams } from '../../types';
import settings from '../../utils/settings';
import { useAppDefinition } from '../AppDefinitionProvider';
import BlockList from '../BlockList';
import FlowPage from '../FlowPage';
import PageDialog from '../PageDialog';
import TabsPage from '../TabsPage';
import TitleBar from '../TitleBar';
import { useUser } from '../UserProvider';
import styles from './index.css';
import messages from './messages';

export default function Page(): ReactElement {
  const { definition } = useAppDefinition();
  const redirect = useLocationString();
  const { isLoggedIn, role } = useUser();
  const {
    params: { lang, pageId },
    path,
    url,
  } = useRouteMatch<{ lang: string; pageId: string }>();

  const [dialog, setDialog] = useState<ShowDialogParams>();

  const ee = useRef<EventEmitter>();
  if (!ee.current) {
    ee.current = new EventEmitter();
  }

  const index = definition.pages.findIndex((p) => normalize(p.name) === pageId);
  const page = index === -1 ? null : definition.pages[index];
  const prefix = index === -1 ? null : `pages.${index}`;

  const showDialog = useCallback((d: ShowDialogParams) => {
    setDialog(d);
    return () => {
      setDialog(null);
    };
  }, []);

  useEffect(() => {
    if (!page) {
      return;
    }
    const queryStringParams = new URLSearchParams({ ...definition.theme, ...page.theme });
    const bulmaStyle = document.getElementById('bulma-style-app') as HTMLLinkElement;
    const bulmaUrl = new URL(bulmaStyle.href);
    queryStringParams.sort();
    bulmaUrl.search = String(queryStringParams);
    bulmaStyle.href = String(bulmaUrl);
  }, [definition, page]);

  useEffect(() => {
    if (ee.current) {
      ee.current.removeAllListeners();
      ee.current = null;
      ee.current = new EventEmitter();
    }

    return () => {
      if (ee.current) {
        ee.current.removeAllListeners();
        ee.current = null;
      }
    };
  }, [page]);

  const checkPagePermissions = (p: PageDefinition): boolean => {
    const roles = p.roles || definition.roles || [];
    return roles.length === 0 || roles.some((r) => checkAppRole(definition.security, r, role));
  };

  // If the user is on an existing page and is allowed to view it, render it.
  if (page && checkPagePermissions(page)) {
    return (
      <main className={styles.root} data-path={prefix}>
        <TitleBar>{page.name}</TitleBar>
        {page.type === 'tabs' ? (
          <TabsPage
            ee={ee.current}
            page={page}
            prefix={prefix}
            showDialog={showDialog}
            subPages={page.subPages}
          />
        ) : (
          // The switch is used to enforce an exact path.
          <Switch>
            <Route exact path={path}>
              {page.type === 'flow' ? (
                <FlowPage
                  definition={definition}
                  ee={ee.current}
                  page={page}
                  prefix={prefix}
                  showDialog={showDialog}
                />
              ) : (
                <BlockList
                  blocks={page.blocks}
                  ee={ee.current}
                  page={page}
                  prefix={`${prefix}.blocks`}
                  showDialog={showDialog}
                />
              )}
            </Route>
            {/* Redirect from a matching sub URL to the actual URL */}
            <Redirect to={url} />
          </Switch>
        )}
        <PageDialog dialog={dialog} ee={ee.current} page={page} showDialog={showDialog} />
      </main>
    );
  }

  // If the user isn’t allowed to view the page, because they aren’t logged in, redirect to the
  // login page.
  if (!isLoggedIn) {
    return <Redirect to={`/${lang}/Login?${new URLSearchParams({ redirect })}`} />;
  }

  // If the user is logged in, but isn’t allowed to view the current page, redirect to the default
  // page.
  const defaultPage = definition.pages.find((p) => p.name === definition.defaultPage);
  if (checkPagePermissions(defaultPage)) {
    return <Redirect to={`/${lang}/${normalize(defaultPage.name)}`} />;
  }

  // If the user isn’t allowed to view the default page either, find a page to redirect the user to.
  const redirectPage = definition.pages.find(checkPagePermissions);
  if (redirectPage) {
    return <Redirect to={`/${lang}/${normalize(redirectPage.name)}`} />;
  }

  // If the user isn’t allowed to view any pages, show an error message.
  return (
    <Content padding>
      <Message color="danger">
        <p>
          <FormattedMessage
            {...messages.permissionError}
            values={{
              a: (text: string) => (
                <a
                  href={`${settings.apiUrl}/apps/${settings.id}`}
                  rel="noopener noreferrer"
                  target="_blank"
                >
                  {text}
                </a>
              ),
            }}
          />
        </p>
        <Button className="mt-4" color="danger">
          <FormattedMessage {...messages.logout} />
        </Button>
      </Message>
    </Content>
  );
}
