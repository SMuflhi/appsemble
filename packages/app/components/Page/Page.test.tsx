import { BasicPage, FlowPage, TabsPage } from '@appsemble/types';
import { shallow } from 'enzyme';
import React from 'react';

import Page from './Page';

jest.mock('react-router-dom', () => ({
  useHistory: () => ({
    push: jest.fn(),
  }),
}));

jest.mock('react-intl', () => {
  const reactIntl = require.requireActual('react-intl');
  const intl = reactIntl.createIntl({
    locale: 'en',
  });

  return {
    ...reactIntl,
    useIntl: () => intl,
  };
});

beforeEach(() => {
  jest.spyOn(document, 'getElementById').mockReturnValue(document.createElement('link'));
});

it('should render the blocks for a page', () => {
  const page: BasicPage = {
    name: 'Test Page',
    blocks: [
      { type: 'test', version: '0.0.0' },
      { type: 'test', version: '0.0.0' },
    ],
  };
  const wrapper = shallow(<Page page={page} />);

  expect(wrapper).toMatchSnapshot();
});

it('should render tabs pages', () => {
  const page: TabsPage = {
    name: 'Test Page',
    type: 'tabs',
    subPages: [
      {
        name: 'Sub A',
        blocks: [
          { type: 'test', version: '0.0.0' },
          { type: 'test', version: '0.0.0' },
        ],
      },
      {
        name: 'Sub B',
        blocks: [
          { type: 'test', version: '0.0.0' },
          { type: 'test', version: '0.0.0' },
        ],
      },
    ],
  };
  const wrapper = shallow(<Page page={page} />);

  expect(wrapper).toMatchSnapshot();
});

it('should render flow page', () => {
  const page: FlowPage = {
    name: 'Test Page',
    type: 'flow',
    subPages: [
      {
        name: 'Sub A',
        blocks: [
          { type: 'test', version: '0.0.0' },
          { type: 'test', version: '0.0.0' },
        ],
      },
      {
        name: 'Sub B',
        blocks: [
          { type: 'test', version: '0.0.0' },
          { type: 'test', version: '0.0.0' },
        ],
      },
    ],
  };
  const wrapper = shallow(<Page page={page} />);

  expect(wrapper).toMatchSnapshot();
});
