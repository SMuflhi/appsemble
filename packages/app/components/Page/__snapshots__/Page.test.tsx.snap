// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render flow page 1`] = `
<Fragment>
  <TitleBar>
    Test Page
  </TitleBar>
  <FlowPage
    blocks={Array []}
    counter={0}
    definition={
      Object {
        "defaultPage": "Test Page",
        "pages": Array [
          Object {
            "name": "Test Page",
            "subPages": Array [
              Object {
                "blocks": Array [
                  Object {
                    "type": "test",
                    "version": "0.0.0",
                  },
                  Object {
                    "type": "test",
                    "version": "0.0.0",
                  },
                ],
                "name": "Sub A",
              },
              Object {
                "blocks": Array [
                  Object {
                    "type": "test",
                    "version": "0.0.0",
                  },
                  Object {
                    "type": "test",
                    "version": "0.0.0",
                  },
                ],
                "name": "Sub B",
              },
            ],
            "type": "flow",
          },
        ],
      }
    }
    page={
      Object {
        "name": "Test Page",
        "subPages": Array [
          Object {
            "blocks": Array [
              Object {
                "type": "test",
                "version": "0.0.0",
              },
              Object {
                "type": "test",
                "version": "0.0.0",
              },
            ],
            "name": "Sub A",
          },
          Object {
            "blocks": Array [
              Object {
                "type": "test",
                "version": "0.0.0",
              },
              Object {
                "type": "test",
                "version": "0.0.0",
              },
            ],
            "name": "Sub B",
          },
        ],
        "type": "flow",
      }
    }
    showDialog={[Function]}
  />
  <Connect(PageDialog) />
</Fragment>
`;

exports[`should render tabs pages 1`] = `
<Fragment>
  <TitleBar>
    Test Page
  </TitleBar>
  <TabsPage
    counter={0}
    showDialog={[Function]}
    subPages={
      Array [
        Object {
          "blocks": Array [
            Object {
              "type": "test",
              "version": "0.0.0",
            },
            Object {
              "type": "test",
              "version": "0.0.0",
            },
          ],
          "name": "Sub A",
        },
        Object {
          "blocks": Array [
            Object {
              "type": "test",
              "version": "0.0.0",
            },
            Object {
              "type": "test",
              "version": "0.0.0",
            },
          ],
          "name": "Sub B",
        },
      ]
    }
  />
  <Connect(PageDialog) />
</Fragment>
`;

exports[`should render the blocks for a page 1`] = `
<Fragment>
  <TitleBar>
    Test Page
  </TitleBar>
  <Connect(BlockList)
    blocks={
      Array [
        Object {
          "type": "test",
          "version": "0.0.0",
        },
        Object {
          "type": "test",
          "version": "0.0.0",
        },
      ]
    }
    counter={0}
    showDialog={[Function]}
  />
  <Connect(PageDialog) />
</Fragment>
`;
